### ISSUE TITLE
Modify NIRSpec IFU level-3 ASN rules to include only one grating

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#4845]
_Issue [JP-1413](https://jira.stsci.edu/browse/JP-1413) was created on JIRA by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse):_

If a NIRSpec IFU (EXP_TYPE='NRS_IFU') observation contains a mixture of exposures using multiple gratings (e.g. g235h, g395h) the current ASN rules create spec3 ASN's that contain the exposures for all gratings. In theory this is acceptable, because there is some wavelength overlap between adjacent gratings and hence merged spectral cubes can be created. However, at least right now there are tremendous performance issues associated with having such large ASN's, especially if multi-point and multi-tile dither/mosaic patterns are in use. There are occasions where such large ASN's have also run into memory errors in Ops processing (e.g. JWSTDMS-44).

It would be prudent, therefore, to at least temporarily modify the ASN rules to create separate spec3 ASN's for each grating value encountered in a NIRSpec IFU observation, to help alleviate the processing performance problems.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1413?focusedCommentId=460309#comment-460309):_

The test data directory includes the original pool file "jw00626_20200307t140611_pool.csv" and the original spec3 ASN file for observation 29 "jw00626-o029_20191129t190337_spec3_001_asn.json". That spec3 ASN contains a total of 16 exposures, 8 that use grating g235h and 8 that use grating g395h.

With the modified ASN rules, 2 spec3 files should be created, with one containing the 8 g235h exposures, and the other containing the 8 g395h exposures. Examples of such spec3 ASN's (created manually) are "jw00626-o029_20191129t190337_spec3_002_asn.json" and "jw00626-o029_20191129t190337_spec3_003_asn.json".

**Comment by stscijgbot-jp** [comment]
_Comment by [David Davis](https://jira.stsci.edu/secure/ViewProfile.jspa?name=ddavis) on [JIRA](https://jira.stsci.edu/browse/JP-1413?focusedCommentId=461655#comment-461655):_

Updated the asn rules to segregate observations based on the grating in use. An example for jw00626-o029 is attached above.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1413?focusedCommentId=462865#comment-462865):_

I think the internal machinery already took care of the product file name issue. For example, in the original spec3 ASN in this example, the product name specified in the ASN file was "jw00626-o029_t006_nirspec_clear". The resulting products then got the grating and filter names appended to this product root, to produce "jw00626-o029_t006_nirspec_clear_g235h-f170lp_s3d.fits" and "jw00626-o029_t006_nirspec_clear_g395h-f290lp_s3d.fits". Although I'm not sure why the "clear" was part of the product name root.

Well, we'll see what happens with the new ASN's when used to run through spec3 processing.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Davis](https://jira.stsci.edu/secure/ViewProfile.jspa?name=ddavis) on [JIRA](https://jira.stsci.edu/browse/JP-1413?focusedCommentId=462873#comment-462873):_

In the original implementation the product name was
```java
"name": "jw00626-o029_t006_nirspec_clear"```
When checking for duplicates the code only checks to see if the product name is repeated in associations. When the associations are split by grating then that needed to be added to the product name to prevent the duplicate checking from tossing one of the new associations. The code certainly does this in other cases.

The optical element is set to "clear" if the optical element string returns a blank string.

An additional complication is that if an observation is a targeted background the processing is through a separate module. I've also added rules to split the nrs_ifu background observations by grating.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1413?focusedCommentId=462888#comment-462888):_

(y)

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-1413?focusedCommentId=464790#comment-464790):_

Fixed in #4926
